### PR TITLE
PERF: avoid copies in lib.infer_dtype

### LINKED ISSUE #45022: API: infer_dtype with skipna=True only skip valid-for-dtype NAs
I'm working on making lib.infer_dtype copy-free and finding things would be easier/more consistent if we tweaked the meaning of the skipna keyword.

In particular, instead of doing `values = values[~isnaobj(values)]`, followed by e.g. `is_string_array(values)`, we could do `is_string_array(values, skipna=skipna)`.  This would change the results in cases where we have NA values that are not considered valid_na by is_string_array, e.g. in the status quo:

```
import pandas as pd
import numpy as np
from pandas._libs import lib

arr = np.array(["foo", pd.NaT, "bar"], dtype=object)

In [2]: lib.infer_dtype(arr, skipna=True)
Out[2]: 'string'

In [3]: lib.is_string_array(arr, skipna=True)
Out[3]: False
```

So the suggestion here is to change [2] to give 'mixed'.  I'm finding that to make this work without breaking the world we also need to change StringValidator.is_valid_null to accept np.nan and None (currently just accepts pd.NA)

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #45022
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

In the `skipna=True` case, instead of doing `values = values[~isnaobj(values)]`, just pass `skipna` to the relevant validator functions.  This causes trouble because those validator functions have different rules about what NA values they allow (analogous to is_valid_na_for_dtype).

I'm down to 1 test case failing locally, fixing it will require changing the StringArray constructor to allow None and np.nan (xref #40839) in addition to just pd.NA (AFAICT pd.array and pd.Series(... dtype="string") will not be affected).  I'm fine with that, but someone more involved with the string code might want to weigh in cc @simonjayhawkins 

```
In [1]: import numpy as np

In [2]: arr = np.arange(10**5).astype(object)

In [3]: from pandas._libs import lib

In [4]: %timeit lib.infer_dtype(arr, skipna=True)
962 µs ± 77 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <- PR
3.81 ms ± 33 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  # <- master


In [7]: %load_ext memory_profiler

In [16]: arr = arr.repeat(100)

In [18]: %memit lib.infer_dtype(arr, skipna=True)
peak memory: 169.24 MiB, increment: 0.00 MiB  # <- PR
peak memory: 265.39 MiB, increment: 95.38 MiB  # <- master

```

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Bad news @mroeschke:

```
    def test_dialect_conflict_except_delimiter(all_parsers, custom_dialect, arg, value):
[...]
E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed file <_io.BufferedRandom name=10>'), 'D:\\a\\1\\s\\pandas\\core\\indexes\\base.py', 665)]
```

**Comment by mroeschke** [comment]
> Bad news @mroeschke:
> 
> ```
>     def test_dialect_conflict_except_delimiter(all_parsers, custom_dialect, arg, value):
> [...]
> E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed file <_io.BufferedRandom name=10>'), 'D:\\a\\1\\s\\pandas\\core\\indexes\\base.py', 665)]
> ```

:(

This looks like it occurred on Windows? The check I added doesn't work there since idk the equivalent `lsof` check.

**Comment by jreback** [comment]
wow, looks pretty good.

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -1882,7 +1889,7 @@ cdef class StringValidator(Validator):
 
     cdef bint is_valid_null(self, object value) except -1:
         # We deliberately exclude None / NaN here since StringArray uses NA
-        return value is C_NA
+        return value is C_NA or value is None or util.is_nan(value)
```
Comment (lithomas1) [inline pandas/_libs/lib.pyx:1892]: Quick question: 
If I do ``pd.arrays.StringArray(np.array(["a", np.nan], dtype=object))``, won't this result in a StringArray with ``np.nan`` in it? (something like this? https://github.com/pandas-dev/pandas/issues/30966#issuecomment-573698662)

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1892]: > won't this result in a StringArray with np.nan in it?

yep.  probably ought to do something about that.  have something min mind?

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1892]: I'm trying to fix this in #41412. This PR is probably stuck in the meantime.

**Comment by jreback** [comment]
this WIP?

**Comment by jbrockmendel** [comment]
> this WIP?

@lithomas1 says this needs to wait on #41412

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: @lithomas1 is this un-blocked?

**Comment by jreback** [comment]
nice!

**Pull request merged**